### ISSUE TITLE
Entries in ecsIndex cache are not updated after they expire once

### ISSUE THREAD

**Issue by liordot** [PowerDNS/pdns#6241]
- Program: Recursor 
 - Issue type: Bug report 

### Short description

When using ecs, entries are only served from cache until they expire one time, after that they are not cached anymore.

### Environment

 - Operating system: Ubuntu 16.04
 - Software version: 4.1.0
 - Software source: Ubuntu Repository

### Steps to reproduce

1. Enable ecs
2. Make a request for a host
3. Wait for the response to expire

### Expected behaviour

Entry should be served from cache
### Actual behaviour

Entry is not cached.
### Other information

It looks like the code here:
https://github.com/PowerDNS/pdns/blob/master/pdns/recursor_cache.cc#L271
Only update the ecsIndex if the entry is not in d_cache.
Here the code only erase the ecsIndex but d_cache is still populated.
https://github.com/PowerDNS/pdns/blob/master/pdns/recursor_cache.cc#L121
The result is that the entry is not in the ecsIndex and the response is not cached.

The code below fixes the issue:

```
if (stored == d_cache.end()) {
    stored = d_cache.insert(CacheEntry(key, CacheEntry::records_t(), auth)).first;
    isNew = true;

  }
    if (ednsmask) {
      auto ecsIndexKey = boost::make_tuple(qname, qt.getCode());
      auto ecsIndex = d_ecsIndex.find(ecsIndexKey);
      if (ecsIndex == d_ecsIndex.end()) {
        ecsIndex = d_ecsIndex.insert(ECSIndexEntry(qname, qt.getCode())).first;
      }
      ecsIndex->addMask(*ednsmask);
    }

```

**Comment by ahupowerdns** [comment]
thanks liordot! We're looking into it.

**Comment by ahupowerdns** [comment]
@liordot - we confirm the issue & that your fix appears to be correct. How can we thank you in the release notes?

**Comment by liordot** [comment]
You're welcome :)

Can you share an ETA on when the next version with this fix will be released?

**Comment by rgacogne** [comment]
The fix proposed in e0dfa1c44bc8a356e9f8c0e35f382c2ca2fff239 looks good to me, thank you again for reporting this!

**Comment by ahupowerdns** [comment]
I think this one can be close, @rgacogne ?
